Add repeat-fu

ideasman42 opened the pull request:
### Brief summary of what the package does

A package to repeat the last action that supports configurable groupings.

Or this could be presented as: `evil-repeat` for other modal editing systems (presets are included for Meow, Emacs for example).
Support for other modal editing systems is relatively easy to support.

This allows (among other things) all actions in "insert" mode to be grouped, it addresses the limitation noted here:
https://github.com/meow-edit/meow/discussions/414

NOTE: This started out as an improvement to [dot-mode](https://melpa.org/#/dot-mode), however the functionality changes enough that I don't think it's practical to attempt to make this an improvement to `dot-mode`.

### Direct link to the package repository

https://codeberg.org/ideasman42/emacs-repeat-fu

### Your association with the package

Maintainer.

### Relevant communications with the upstream package maintainer

**None needed**

### Checklist

<!-- Please confirm by replacing `[]` with `[x]`: -->

- [x] The package is released under a [GPL-Compatible Free Software License](https://www.gnu.org/licenses/license-list.en.html#GPLCompatibleLicenses)
- [x] I've read [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)
- [x] I've used the latest version of [package-lint](https://github.com/purcell/package-lint) to check for packaging issues, and addressed its feedback
- [x] My elisp byte-compiles cleanly
- [x] I've used `M-x checkdoc` to check the package's documentation strings
- [x] I've built and installed the package using the instructions in [CONTRIBUTING.org](https://github.com/melpa/melpa/blob/master/CONTRIBUTING.org)

<!-- After submitting, please fix any problems the CI reports. -->

riscy commented:
Thanks for this.  Quick first pass -

⸺ `repeat-fu-preset-single.el` with byte-compile using Emacs 30.1:
```
repeat-fu-preset-single.el:50:22: Warning: the function `repeat-fu-command-test-skip-change' might not be defined at runtime.
repeat-fu-preset-single.el:26:6: Warning: the function `repeat-fu-command-test-skip' might not be defined at runtime.
```

⸺ `repeat-fu-preset-meow.el` with byte-compile using Emacs 30.1:
```
repeat-fu-preset-meow.el:28:2: Error: Cannot open load file: No such file or directory, meow
```

⸺ `repeat-fu-preset-bray.el` with byte-compile using Emacs 30.1:
```
repeat-fu-preset-bray.el:13:2: Error: Cannot open load file: No such file or directory, bray
```

⸺ `repeat-fu-preset-single.el` with package-lint 20250307.912 and `package-lint-main-file` = "repeat-fu.el":
```
1 issue found:
4:0: error: Package-Requires outside the main file have no effect.
```

⸺ `repeat-fu-preset-multi.el` with package-lint 20250307.912 and `package-lint-main-file` = "repeat-fu.el":
```
1 issue found:
4:0: error: Package-Requires outside the main file have no effect.
```

⸺ `repeat-fu-preset-meow.el` with package-lint 20250307.912 and `package-lint-main-file` = "repeat-fu.el":
```
1 issue found:
4:0: error: Package-Requires outside the main file have no effect.
```

⸺ `repeat-fu-preset-bray.el` with package-lint 20250307.912 and `package-lint-main-file` = "repeat-fu.el":
```
1 issue found:
4:0: error: Package-Requires outside the main file have no effect.
```

⸺ `repeat-fu.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- repeat-fu.el#L352: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L353: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L354: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L355: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L343: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L344: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L345: It's safer to sharp-quote function names; use `#'`
- repeat-fu.el#L346: It's safer to sharp-quote function names; use `#'`
```

⸺ `repeat-fu-preset-multi.el` with [melpazoid](https://github.com/riscy/melpazoid):
```
- repeat-fu-preset-multi.el#L91: You may want to consider `(derived-mode-p 'xyz)`
```

⸺ Package and license:
- Add a GPL-compatible LICENSE file to the repository
- repeat-fu-preset-bray.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)
- repeat-fu-preset-meow.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)
- repeat-fu-preset-multi.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)
- repeat-fu-preset-single.el needs *formal* license boilerplate and/or an [SPDX-License-Identifier](https://spdx.dev/ids/)

riscy commented:
> This only depends on an "emacs" version, if this is removed there is a warning that some features depend on newer emacs version.

I think the warning you describe must come from package-lint, so if you set the local variable `package-lint-main-file` within those support files you'll be good to go on that.

> That is the presets files: repeat-fu-preset-*.el which depend on repeat-fu.el. In practice it works for them to (require 'repeat-fu) when used in the package because repeat-fu.el is in the load-path. So I'm not sure if/how that should be addressed.

This is only happening in one file now:

⸺ `repeat-fu-preset-single.el` with byte-compile using Emacs 30.1:
```
repeat-fu-preset-single.el:55:22: Warning: the function `repeat-fu-command-test-skip-change' might not be defined at runtime.
repeat-fu-preset-single.el:31:6: Warning: the function `repeat-fu-command-test-skip' might not be defined at runtime.
```
But why wrap `(eval-when-compile (require 'repeat-fu))` in `eval-when-compile`?  That's the source of the warnings.

ideasman42 commented:
@riscy updated.
- replaced emacs-version with `package-lint-main-file` local variable.o
- Removed `(eval-when-compile ...)`  around `(require 'repeat-fu)`, _thought this was a common convention/practice._
